File: cockroachdb/golang/services/test1_service.go
package services

import (
	"context"
	"database/sql"
	"fmt"
	modelentities "note-golang-cockroachdb/models/entities"
	modelrequests "note-golang-cockroachdb/models/requests"
	modelresponses "note-golang-cockroachdb/models/responses"
	"note-golang-cockroachdb/repositories"
	"note-golang-cockroachdb/utils"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
)

type Test1Service interface {
	Create(ctx context.Context, createRequest modelrequests.CreateRequest) (response modelresponses.Response)
	GetById(ctx context.Context, id string) (response modelresponses.Response)
	GetAll(ctx context.Context) (response modelresponses.Response)
	Update(ctx context.Context, updateRequest modelrequests.UpdateRequest) (response modelresponses.Response)
	Delete(ctx context.Context, deleteRequest modelrequests.DeleteRequest) (response modelresponses.Response)
}

type test1Service struct {
	cockroachDbUtil utils.CockroachDbUtil
	test1Repository repositories.Test1Repository
}

func NewTest1Service(cockroachDbUtil utils.CockroachDbUtil, test1Repository repositories.Test1Repository) Test1Service {
	return &test1Service{
		cockroachDbUtil: cockroachDbUtil,
		test1Repository: test1Repository,
	}
}

func (service *test1Service) Create(ctx context.Context, createRequest modelrequests.CreateRequest) (response modelresponses.Response) {
	tx, err := service.cockroachDbUtil.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}

	defer func() {
		errCommitOrRollback := service.cockroachDbUtil.CommitOrRollback(tx, ctx, err)
		if errCommitOrRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	var test1 modelentities.Test1
	uuidv7, err := uuid.NewV7()
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	test1.Id = uuid.NullUUID{Valid: true, UUID: uuidv7}
	test1.Test = sql.NullString{Valid: true, String: createRequest.Test}
	rowsAffected, err := service.test1Repository.Create(tx, ctx, test1)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	fmt.Println("rowsAffected:", rowsAffected)
	return modelresponses.SetCreateResponse(test1)
}

func (service *test1Service) GetById(ctx context.Context, id string) (response modelresponses.Response) {
	uuidv7, err := uuid.Parse(id)
	if err != nil {
		fmt.Println("err parse uuid getbyid service:", id, err)
		return modelresponses.SetInternalServerErrorResponse()
	}
	test1, err := service.test1Repository.GetById(service.cockroachDbUtil.GetDb(), ctx, uuidv7)
	if err != nil {
		fmt.Println("err getbyid service:", err)
		return modelresponses.SetInternalServerErrorResponse()
	}
	return modelresponses.SetOkResponse(test1)
}

func (service *test1Service) GetAll(ctx context.Context) (response modelresponses.Response) {
	test1s, err := service.test1Repository.GetAll(service.cockroachDbUtil.GetDb(), ctx)
	if err != nil {
		fmt.Println("err getall service:", err)
		return modelresponses.SetInternalServerErrorResponse()
	}
	return modelresponses.SetOkResponse(test1s)
}

func (service *test1Service) Update(ctx context.Context, updateRequest modelrequests.UpdateRequest) (response modelresponses.Response) {
	tx, err := service.cockroachDbUtil.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	defer func() {
		errCommitOrRollback := service.cockroachDbUtil.CommitOrRollback(tx, ctx, err)
		if errCommitOrRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	var test1 modelentities.Test1
	test1.Id = uuid.NullUUID{Valid: true, UUID: updateRequest.Id}
	test1.Test = sql.NullString{Valid: true, String: updateRequest.Test}
	rowsAffected, err := service.test1Repository.Update(tx, ctx, test1)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	fmt.Println("rowsAffected:", rowsAffected)
	return modelresponses.SetOkResponse(test1)
}

func (service *test1Service) Delete(ctx context.Context, deleteRequest modelrequests.DeleteRequest) (response modelresponses.Response) {
	tx, err := service.cockroachDbUtil.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}

	defer func() {
		errCommitRollback := service.cockroachDbUtil.CommitOrRollback(tx, ctx, err)
		if errCommitRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	rowsAffected, err := service.test1Repository.Delete(tx, ctx, deleteRequest.Id)
	if err != nil {
		fmt.Println("err:", deleteRequest.Id, err)
		return modelresponses.SetInternalServerErrorResponse()
	}
	fmt.Println("rowsAffected:", rowsAffected)
	return modelresponses.SetNoContentResponse()
}

File: cockroachdb/rust/src/services/test1_service.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::{models::{entities::test1::Test1, requests::{create_request::CreateRequest, delete_request::DeleteRequest, update_request::UpdateRequest}, responses::{create_response::CreateResponse, get_all_response::GetAllResponse, get_by_id_response::GetByIdResponse, response::Response, update_response::UpdateResponse}}, repositories::test1_repository::{Test1Repository, Test1RepositoryImpl}, utils::cockroachdb_util::{CockroachDbUtil, CockroachDbUtilImpl}};

pub trait Test1Service {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse>;
    async fn get_by_id(&self, id: Uuid) -> Response<GetByIdResponse>;
    async fn get_all(&self) -> Response<Vec<GetAllResponse>>;
    async fn update(&self, update_request: UpdateRequest) -> Response<UpdateResponse>;
    async fn delete(&self, delete_request: DeleteRequest) -> Response<()>;
}

pub struct Test1ServiceImpl {
    cockroachdb_util: Arc<CockroachDbUtilImpl>,
    test1_repository: Arc<Test1RepositoryImpl>
}

impl Test1ServiceImpl {
    pub fn new(cockroachdb_util: Arc<CockroachDbUtilImpl>, test1_repository: Arc<Test1RepositoryImpl>) -> Test1ServiceImpl {
        Test1ServiceImpl { 
            cockroachdb_util,
            test1_repository
         }
    }
}

impl Test1Service for Test1ServiceImpl {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse> {
        let mut tx = match self.cockroachdb_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response()
            }
        };

        let test1 = Test1 {id: Uuid::now_v7(), test: create_request.test};
        let rows_affected = match self.test1_repository.create(&mut tx, &test1).await {
            Ok(id) => id,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };
        if rows_affected != 1 {
            return Response::set_internal_server_error_response();
        }

        match self.cockroachdb_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }
        return Response::set_created_response(CreateResponse::set_craeate_response(test1));
    }

    async fn get_by_id(&self, id: Uuid) -> Response<GetByIdResponse> {
        let test1 = match self.test1_repository.get_by_id(self.cockroachdb_util.get_pool().await, id).await {
            Ok(test1) => test1,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            } 
        };
        return Response::set_ok_response(GetByIdResponse::set_get_by_id_response(test1));
    }

    async fn get_all(&self) -> Response<Vec<GetAllResponse>> {
        let test1s = match self.test1_repository.get_all(self.cockroachdb_util.get_pool().await).await {
            Ok(test1s) => test1s,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            } 
        };
        return Response::set_ok_response(GetAllResponse::set_get_all_response(test1s));
    }

    async fn update(&self, update_request: UpdateRequest) -> Response<UpdateResponse> {
        let mut tx = match self.cockroachdb_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        let test1 = Test1 {id: update_request.id, test: update_request.test};
        let rows_affected = match self.test1_repository.update(&mut tx, &test1).await {
            Ok(rows_affected) => rows_affected,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        if rows_affected != 1 {
            return Response::set_internal_server_error_response();
        }

        match self.cockroachdb_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }

        return Response::set_ok_response(UpdateResponse::set_update_response(test1))
    }

    async fn delete(&self, delete_request: DeleteRequest) -> Response<()> {
        let mut tx = match self.cockroachdb_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };
        
        let rows_affected = match self.test1_repository.delete(&mut tx, delete_request.id).await {
            Ok(rows_affected) => rows_affected,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        if rows_affected != 1 {
            return Response::set_internal_server_error_response();
        }

        match self.cockroachdb_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }

        return Response::set_no_content_response()
    }
}